pyteal/stack_frame.py

```diff
@@ -0,0 +1,701 @@
+    FLAGGED_BY_DEV = "Developer has flagged expression as compiler generated"
+
+
+_PT_GEN = {
+    "# T2PT0": PT_GENERATED.PRAGMA,
+    "# T2PT1": PT_GENERATED.SUBR_LABEL,
+    "# T2PT2": PT_GENERATED.RETURN_NONE,
+    "# T2PT3": PT_GENERATED.RETURN_VALUE,
+    "# T2PT4": PT_GENERATED.SUBR_PARAM,
+    "# T2PT5": PT_GENERATED.BRANCH,
+    "# T2PT6": PT_GENERATED.BRANCH_LABEL,
+    "# T2PT7": PT_GENERATED.TYPE_ENUM_TXN,
+    "# T2PT8": PT_GENERATED.TYPE_ENUM_ONCOMPLETE,
+}
+
+
+class PytealFrameStatus(IntEnum):
+    """integer values indicate 'confidence' on a scale of 0 - 10"""
+
+    MISSING = 0
+    MISSING_AST = 1
+    MISSING_CODE = 2
+    PYTEAL_GENERATED = 3
+    PATCHED_BY_PREV_OVERRIDE_NEXT = 4
+    PATCHED_BY_NEXT_OVERRIDE_PREV = 5
+    PATCHED_BY_PREV = 6
+    PATCHED_BY_NEXT = 7
+    PATCHED_BY_PREV_AND_NEXT = 8
+    COPACETIC = 9  # currently, 90% confident is as good as it gets
+
+    def human(self) -> str:
+        match self:
+            case self.MISSING:
+                return "sourcemapping line: total failure"
+            case self.MISSING_AST:
+                return "unreliable as missing AST"
+            case self.MISSING_CODE:
+                return "unreliable as couldn't retrieve source"
+            case self.PYTEAL_GENERATED:
+                return "INCOMPLETE"  # "incomplete as source not user generated"
+            case self.PATCHED_BY_PREV_OVERRIDE_NEXT:
+                return "previously INCOMPLETE- patched to previous frame (ambiguous)"
+            case self.PATCHED_BY_NEXT_OVERRIDE_PREV:
+                return "previously INCOMPLETE- patched to next frame (ambiguous)"
+            case self.PATCHED_BY_PREV:
+                return "previously INCOMPLETE- patched to previous frame"
+            case self.PATCHED_BY_NEXT:
+                return "previously INCOMPLETE- patched to next frame"
+            case self.PATCHED_BY_PREV_AND_NEXT:
+                return "previously INCOMPLETE- patched"
+            case self.COPACETIC:
+                return "COPACETIC"
+
+        raise Exception(f"unrecognized {type(self)} - THIS SHOULD NEVER HAPPEN!")
+
+
+class PyTealFrame(StackFrame):
+    """
+    PyTealFrame is the middle generation in the StackFrame class hierarchy.
+
+    It adds a richer and more polished set of methods to learn about the PytTeal source.
+    """
+
+    def __init__(
+        self,
+        frame_info: FrameInfo,
+        node: AST | None,
+        creator: NatalStackFrame,
+        full_stack: list[FrameInfo] | None,
+        rel_paths: bool = True,
+        parent: Optional["PyTealFrame"] | None = None,
+    ):
+        super().__init__(frame_info, node, creator, full_stack)
+        self.rel_paths = rel_paths
+        self.parent = parent
+
+        self._raw_code: str | None = None
+        self._status: PytealFrameStatus | None = None
+
+    def __repr__(self) -> str:
+        """TODO: this repr isn't compliant. Should we keep it anyway for convenience?"""
```

Comment:
maybe call it something else

Reply:
Pushing back, because as misleading as it is, it's handy when debugging in VS-code. Shall I change the comment to say instead of "TODO" to explain the non-compliance?

Reply:
[85a633e](https://github.com/algorand/pyteal/pull/650/commits/85a633ebf6f99dfff8f15cc968af962a2335417f)